[Pipelining] Reorder _Action from F1_1 to 1F1

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129810
* __->__ #129786

Also steers away from accesing _Action via positional unpacking since
that is error prone

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @chauhang @d4l3k

wconstab committed:
Update

[ghstack-poisoned]

awgu commented:
Just so we do not forget: We need to experiment with/evaluate whether it is okay or not that if we run `unshard(async_op=True)` today, the all-gather copy-in ops will run in the current/default stream, so they cannot overlap with compute. If we need to allow this, then we need some changes or some other API, but there is some nuance there (I can explain later).

wconstab committed:
Update

[ghstack-poisoned]

kwen2501 commented:
Can you share more context on the need of adding UNSHARD/RESHARD actions? Is this a more advanced form of pipeline + FSDP so that people can have finer grain control or better overlap?

awgu commented:
@kwen2501 yes :)

H-Huang committed:
Update on "[Pipelining] add UNSHARD/RESHARD actions and test helpers"


Also steers away from accesing _Action via positional unpacking since
that is error prone

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l chauhang d4l3k

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -30,15 +30,20 @@
             _ComputationType.FORWARD: "F",
             _ComputationType.BACKWARD: "B",
             _ComputationType.WEIGHT: "W",
+            _ComputationType.UNSHARD: "UNSHARD",
```
i almost did that but i decided against.
* F,B,W is pretty easy to follow and its nice to have single-letters for the simple layer of the IR
* but once we add a lot of things we'll start hitting conflicts.  I plan to add Send/Recv next, and R would be recv or reshard.  I think for the lower layer of IR it is better to not try to be so brief.

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented:
FYI I split the original PR so this one is just focusing on the refactor of _Action class.  Then i add UNSHARD/RESHARD stuff in the next PR.

wconstab commented:
@pytorchbot merge
